File: mobile/ios/FitnessAI/Uploader.swift
import Foundation

/// Posts readings to the backend /ingest endpoint using the Keychain pairing token.
/// HTTPS is enforced by App Transport Security (no exceptions in Info.plist).
/// No analysis happens on-device — the phone only ships raw readings.
enum Uploader {

    struct Reading: Codable {
        let timestamp: String   // ISO-8601
        var steps: Int?
        var heart_rate: Double?
        var distance_meters: Double?
        var active_calories: Double?
    }

    /// Send a batch of readings. Returns true on success (HTTP 2xx).
    static func upload(readings: [Reading]) async -> Bool {
        guard let token = Keychain.token(), !readings.isEmpty else { return false }
        guard let url = URL(string: "\(Config.backendURL)/ingest/") else { return false }

        let body: [String: Any] = [
            "device": "fitness_phone_ios",
            "app_version": Config.appVersion,
            "readings": readings.map { r in
                var d: [String: Any] = ["timestamp": r.timestamp]
                if let v = r.steps { d["steps"] = v }
                if let v = r.heart_rate { d["heart_rate"] = v }
                if let v = r.distance_meters { d["distance_meters"] = v }
                if let v = r.active_calories { d["active_calories"] = v }
                return d
            },
            "workouts": [],
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: body) else { return false }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.httpBody = data
        req.timeoutInterval = 20

        do {
            let (_, resp) = try await URLSession.shared.data(for: req)
            return (resp as? HTTPURLResponse).map { (200..<300).contains($0.statusCode) } ?? false
        } catch {
            // TODO: persist failed batches to disk and retry (offline queue), like Android's Prefs.enqueue.
            return false
        }
    }
}

File: mobile/android/app/src/main/java/studio/tin/fitnessai/Uploader.kt
package studio.tin.fitnessai

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Posts tracked data to the backend `/ingest` endpoint using the pairing token.
 * Offline-first: if the network is down (or the POST fails), the payload is queued
 * in encrypted prefs and retried on the next successful upload.
 *
 * No analysis happens here — the phone only ships raw readings/workouts; the server
 * stores them in Supabase under the account and analyzes on request.
 */
object Uploader {

    /** Build the /ingest body from collected readings. See models/watch.py for the schema. */
    fun buildPayload(steps: Int, route: List<DoubleArray>?, isoTimestamp: String): String {
        val reading = JSONObject()
            .put("timestamp", isoTimestamp)
            .put("steps", steps)
        val readings = JSONArray().put(reading)
        val body = JSONObject()
            .put("device", "fitness_phone_android")
            .put("app_version", BuildConfig.VERSION_NAME)
            .put("readings", readings)
            .put("workouts", JSONArray())
        // TODO: when a workout/route is active, also append a workout object + post the
        //       route to /routes (lat/lng array) so distance/pace/calories are computed.
        return body.toString()
    }

    /** Try to flush any queued payloads, then send [payloadJson]. Returns true if [payloadJson] sent. */
    fun upload(ctx: Context, payloadJson: String): Boolean {
        val token = Prefs.token(ctx) ?: return false.also { Prefs.enqueue(ctx, payloadJson) }

        // flush backlog first (best-effort)
        for (queued in Prefs.drainQueue(ctx)) {
            if (!post(token, queued)) { Prefs.enqueue(ctx, queued); Prefs.enqueue(ctx, payloadJson); return false }
        }
        val ok = post(token, payloadJson)
        if (!ok) Prefs.enqueue(ctx, payloadJson)
        return ok
    }

    private fun post(token: String, body: String): Boolean {
        return try {
            val conn = (URL("${BuildConfig.BACKEND_URL}/ingest/").openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = 15000
                readTimeout = 20000
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("Authorization", "Bearer $token")
            }
            conn.outputStream.use { it.write(body.toByteArray()) }
            val code = conn.responseCode
            conn.disconnect()
            code in 200..299
        } catch (e: Exception) {
            false
        }
    }
}
